### PR TITLE
Download tokens

### PR DESCRIPTION
**Pull request by ffont**
See description in issue https://github.com/MTG/freesound/issues/796

### CODE REVIEW

File: apiv2/models.py [hunk 1]
```diff
@@ -105,3 +102,20 @@ def client_secret(self):
     @property
     def version(self):
         return "V2"
+
+
+class DownloadToken(models.Model):
```
Comment (bdejong) [inline apiv2/models.py]: Why not using something like JWT tokens for this? That way you don't need to store anything in the DB!

As the length of the token in this case is irrelevant you can dump all the information you need in the token JSON.

Sign it with a private key and you can check the signature when you receive the token again.

File: apiv2/models.py [hunk 2]
```diff
@@ -105,3 +102,20 @@ def client_secret(self):
...
+    client_id = models.ForeignKey(ApiV2Client, null=True)
+    token = models.CharField(max_length=255, db_index=True)
+    expires = models.DateTimeField()
+    created = models.DateTimeField(auto_now_add=True)
+    sound = models.ForeignKey('sounds.Sound')
+
+    def is_expired(self):
+        if not self.expires:
+            return True
+        return timezone.now() >= self.expires
```
Comment (bdejong) [inline apiv2/models.py]: This is an inherent property of JWT tokens: they have a date validity.

**Comment by ffont** [comment]
Hmm sure, makes total sense, I guess I did not know about JWT and I didn't think about this option.
I'll have a look. Any recommendation for a python library? I see there are a number...

**Comment by bdejong** [comment]
https://github.com/jpadilla/pyjwt/

**Comment by alastair** [comment]
We use a home-made variation of this for delete sound links (and maybe some other things too). If we think that it's a good idea, should we use the same tool in all places in FS?

File: apiv2/views.py [hunk 3]
```diff
@@ -983,6 +1003,27 @@ def post(self, request,  *args, **kwargs):
...
+@permission_classes([])
+def download_from_token(request, token):
+    try:
+        token_contents = jwt.decode(token, settings.SECRET_KEY, leeway=10)
+    except jwt.ExpiredSignatureError:
+        raise UnauthorizedException(msg="This token has expried.")
+    except jwt.InvalidTokenError:
+        raise UnauthorizedException(msg="Invalid token.")
+    try:
+        sound = Sound.objects.get(id=token_contents.get('sound_id', None))
```
Comment (alastair) [inline apiv2/views.py]: single or double quotes?

**Comment by alastair** [comment]
Looks OK to me. We talked in person about if we wanted to log a download to the downloads table (not now)

We should consider also putting in the AC spec that a download token should be submitted as a header too, so that services who don't want to read the parameter from the URL have a way of doing auth

File: apiv2/views.py [hunk 4]
```diff
@@ -54,14 +54,14 @@
     from collections import OrderedDict
 except:
     from freesound.utils.ordered_dict import OrderedDict
-from urllib import unquote, quote
+from urllib import quote
```
Comment (bdejong) [inline apiv2/views.py]: unneeded import

File: apiv2/views.py (hunk 3 above)
Comment (bdejong) [inline apiv2/views.py]: the last rules I read were: always use single everywhere, except if your string contains a single quote, then use double. :-)

File: apiv2/views.py [hunk 5]
```diff
@@ -482,6 +482,26 @@ def get(self, request,  *args, **kwargs):
...
+    __doc__ = 'Get a url to download a sound without authentication.'
+
+    def get(self, request, *args, **kwargs):
+        sound_id = kwargs['pk']
+        logger.info(self.log_message('sound:%i get_download_link' % (int(sound_id))))
+        try:
+            sound = Sound.objects.get(id=sound_id, moderation_state="OK", processing_state="OK")
+        except Sound.DoesNotExist:
+            raise NotFoundException(resource=self)
+        download_token = jwt.encode({
```
Comment (bdejong) [inline apiv2/views.py]: some newlines to separate logical blocks would be nice ;)

File: apiv2/models.py [hunk 6]
```diff
@@ -24,10 +24,7 @@
 from django.conf import settings
 from django.core.urlresolvers import reverse
 from django.contrib.sites.models import Site
-
-
-# Monkeypatch __unicode__ in provider.oauth2.models.Client class as it is not very informative (and it shows better in the admin)
-#Client.__unicode__ = lambda self: u'%s, %s, %s' % (self.name, self.user, self.client_id) #new_unicode
+from django.utils import timezone
```
Comment (bdejong) [inline apiv2/models.py]: Unused import

File: apiv2/views.py (hunk 4 above)
Comment (ffont) [inline apiv2/views.py]: `quote`? `quote` is used. I removed `unquote` because pycharm highlighted it as not used, but in fact these lines are unrelated with the changes in the PR, is one of these code "tidy ups" that I should have probably committed separately but that, hmm, not too bad is it? ;)

File: apiv2/views.py (hunk 3 above)
Comment (ffont) [inline apiv2/views.py]: sure, never got to be consistent with this. I always try to use single quote, then some strings come which have single quotes inside and I try to continue using single quote and backslash the quotes inside. Then  I got tired and I switch to double quote again. Then I just forget about what did I decide...
